### ISSUE TITLE
[Feature Request] Rename `[Unsafe]Pointer.address_of(x)` to something shorter

### ISSUE THREAD

**Issue by soraros** [modular/max#3602]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?

As title. Maybe to `Pointer(to=thing)`.

### What is your motivation for this change?

- I think they take too much line space
- Quote from Chris: "In general, in my opinion, we should use initializers to create instances of a type, not static methods."
- Again, from Chris: "so long as `UnsafePointer` and `Pointer` are consistent, I am happy"

### Any other details?

For the bike-shading crowd: I mostly just moved Chris's comments from Discord, please don't shoot the passenger.

**Comment by melodyogonna** [comment]
+1 to this, conveys the same meaning with a more succinct syntax, definitely more Pythonic.

**Comment by lattner** [comment]
I don't think that making it shorter is the only motivation, it is also about consistency:  I'd suggest using initializer syntax, so that "creating a T" is always done with initializers.

If we did that, it would argue for something like `Pointer(to=x)` or `Pointer(at=x)` or `Pointer(of=x)` or `Pointer(address=x)` etc.

**Comment by martinvuyk** [comment]
Yesss, this will even spare having to set the type that the pointer is for since it can be inferred :+1: 

> Pointer(at=x)  [...]  Pointer(address=x)

these ones in particular might be confusing if you pass an Int and you come from the C world though

**Comment by martinvuyk** [comment]
Opened PR #3606 but I'm getting a segfault for `_ListIter` even though I only added a constructor for `Pointer` and didn't touch any more code :/

**Comment by soraros** [comment]
> If we did that, it would argue for something like `Pointer(to=x)` or `Pointer(at=x)` or `Pointer(of=x)` or `Pointer(address=x)` etc.

More broadly speaking, the fact that Mojo supports function overloads opens the door to Swift-style, argument-label-rich API designs. I'm overall supportive to this approach, but we also need to aware of the trade-offs.

For object initialisation, static methods (technically `classmethods`) are like a "poor man's" overloading in Python. They do have some advantages though:
- They are more "Pythonic", at least in the sense that Python allows this approach (e.g., don't require hacks like those used in `range` to support 3 different arities).
- This is likely not the case for `Pointer`, but it could be important for Python interoperability for some other types (especially when called *from* Python).
- They also have the advantage of making the main argument not keyword-only, which makes them more compatible with higher-order functions.
- They require less LSP support. The LSP doesn't show argument labels in auto-completion, and since it's part of an overload set, the LSP doesn't display the correct docstring, making them even more inaccessible.

**Comment by lattner** [comment]
Awesome, +1 for using initializers to create values :)
